core/src/main/java/jenkins/model/NodeListener.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) Red Hat, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model;
+
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.Node;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+/**
+ * Listen to {@link Node} CRUD operations.
+ *
+ * @author ogondza.
+ * @since TODO
+ */
+public abstract class NodeListener implements ExtensionPoint {
+
+    protected void onCreated(@Nonnull Node node) {}
+
+    protected void onUpdated(@Nonnull Node oldOne, @Nonnull Node newOne) {}
+
+    protected void onDeleted(@Nonnull Node node) {}
+
+    public static void fireOnCreated(@Nonnull Node node) {
+        for (NodeListener nl: all()) {
+            nl.onCreated(node);
```

Comment:
Best practice: Handle RuntimeExceptions coming from implementations, suppress and log them. Otherwise one plugin can break all other instances && node replacement flow